pkg/build/webhook/generic/generic.go

```diff
@@ -42,68 +42,71 @@ func (p *WebHookPlugin) Extract(buildCfg *api.BuildConfig, secret, path string,
+		return revision, envvars, true, warning
+	}
 
-		var data api.GenericWebHookEvent
-		if contentType == "application/yaml" {
-			body, err = yaml.ToJSON(body)
-			if err != nil {
-				glog.V(4).Infof("Error converting payload to json %v, but continuing with build", err)
-				return nil, envvars, true, nil
-			}
-		}
-		if err = json.Unmarshal(body, &data); err != nil {
-			glog.V(4).Infof("Error unmarshalling payload %v, but continuing with build", err)
-			return nil, envvars, true, nil
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return revision, envvars, false, err
+	}
+
+	if len(body) == 0 {
+		return revision, envvars, true, nil
+	}
+
+	var data api.GenericWebHookEvent
+	if contentType == "application/yaml" {
+		body, err = yaml.ToJSON(body)
+		if err != nil {
+			warning := webhook.NewWarning(fmt.Sprintf("error converting payload to json: %v, ignoring payload and continuing with build", err))
+			return revision, envvars, true, warning
 		}
-		if len(data.Env) > 0 && trigger.AllowEnv {
+	}
+	if err = json.Unmarshal(body, &data); err != nil {
+		warning := webhook.NewWarning(fmt.Sprintf("error unmarshalling payload: %v, ignoring payload and continuing with build", err))
+		return revision, envvars, true, warning
+	}
+	if len(data.Env) > 0 {
+		if trigger.AllowEnv {
 			envvars = data.Env
 		}
-		if data.Git == nil {
-			glog.V(4).Infof("No git information for the generic webhook found in %s/%s", buildCfg.Namespace, buildCfg.Name)
-			return nil, envvars, true, nil
-		}
+	} else if data.Git == nil {
+		warning := webhook.NewWarning("no git or env information found in payload, ignoring and continuing with build")
+		return revision, envvars, true, warning
+	}
 
-		if data.Git.Refs != nil {
-			for _, ref := range data.Git.Refs {
-				if webhook.GitRefMatches(ref.Ref, webhook.DefaultConfigRef, &buildCfg.Spec.Source) {
-					revision = &api.SourceRevision{
-						Git: &ref.GitSourceRevision,
-					}
-					return revision, envvars, true, nil
+	if data.Git.Refs != nil {
+		for _, ref := range data.Git.Refs {
+			if webhook.GitRefMatches(ref.Ref, webhook.DefaultConfigRef, &buildCfg.Spec.Source) {
+				revision = &api.SourceRevision{
+					Git: &ref.GitSourceRevision,
 				}
+				return revision, envvars, true, nil
 			}
-			glog.V(2).Infof("Skipping build for BuildConfig %s/%s. None of the supplied refs matched %q", buildCfg.Namespace, buildCfg, buildCfg.Spec.Source.Git.Ref)
-			return nil, envvars, false, nil
-		}
-		if !webhook.GitRefMatches(data.Git.Ref, webhook.DefaultConfigRef, &buildCfg.Spec.Source) {
-			glog.V(2).Infof("Skipping build for BuildConfig %s/%s. Branch reference from %q does not match configuration", buildCfg.Namespace, buildCfg.Name, data.Git.Ref)
-			return nil, envvars, false, nil
-		}
-		revision = &api.SourceRevision{
-			Git: &data.Git.GitSourceRevision,
 		}
+		warning := webhook.NewWarning(fmt.Sprintf("skipping build. None of the supplied refs matched %q", buildCfg.Spec.Source.Git.Ref))
```

Comment:
hm.  that was probably the wrong choice when we implemented the api, but ok I guess we're stuck with it.  Still feels awkward to call it a warning since in all the other cases a warning means "something went wrong but we're proceeding with the build" and here the warning means "something went wrong and we're not doing the build even though we gave you back a 200"

I wonder if we could get away w/ changing this to a non-200 return code even though it's technically breaking the existing api.  @smarterclayton ?